Doc 52 connect to data overview

Rachel-Reverie opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Create the Connect to Data: Overview page
- Restructure the ToC to include Step 2: Connect to Data


After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-


### Definition of Done
Please delete options that are not relevant.

- [X] I have made corresponding changes to the documentation



Thank you for submitting!

Rachel-Reverie committed:
Partial rough draft of overview.md for the Connect to Data step.

Rachel-Reverie committed:
Partial rough draft of overview.md for the Connect to Data step.

Rachel-Reverie committed:
More fleshed out rough draft of overview.md

Rachel-Reverie committed:
More fleshed out rough draft of overview.md

Rachel-Reverie committed:
complete rough draft for overview.md

Rachel-Reverie committed:
complete rough draft for overview.md

Rachel-Reverie committed:
rebased on parent branch

Rachel-Reverie committed:
Minor edits to overview.md
Added overview.md and corresponding how to documents into the ToC

Rachel-Reverie committed:
Merge branch 'universal-map-docs-update' into DOC-52_connect_to_data_overview

Rachel-Reverie committed:
Nested Connect to Data guides under How to guides in the ToC

Rachel-Reverie committed:
Starting updates on index pages for the Connect to Data how to documents.

Rachel-Reverie committed:
Merge remote-tracking branch 'origin/DOC-52_connect_to_data_overview' into DOC-52_connect_to_data_overview

# Conflicts:
#	sidebars.js

talagluck commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+<!-- Brief outline of what the process entails.  -->
+
+Connecting to your data is built around the Datasource object.  A Datasource provides a standard API for accessing and interacting with data from a wide variety of source systems.  It does this by providing an interface for a Data Connector and an Execution Engine to work together.
```
Should we throw in a quick description about what Data Connectors and Execution Engines do here? Otherwise the reference might be confusing.

abegong replied:
These will eventually have tooltips and links to Core Concepts, correct? If so, I think that'd be enough.

Rachel-Reverie replied:
I agree with Don.  And those tags should be included very soon -- I may even add them to this PR just to address this question, and so they are available to use in future documentation as it is composed.

talagluck commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+Please note that this is just a broad outline of the configuration you will be making.  You will find much more detailed examples in our documentation on how to connect to specific source data systems.
+
+The `name` and `class_name` top level keys will be the first you need to define.  The `name` key can be anything you want, but it is best to use a descriptive name as you will use this to reference your Datasource in the future. Unless you are extending Great Expectations and using a subclass of Datasource, you will almost never need to use a `class_name` other than `Datasource` for the top level ...
```
I don't know what our current state is with `SimpleSqlAlchemyDatasource`, so you may need to confirm this with @alexsherstinsky or @donaldheppner.

abegong replied:
Yeah, I suspect that `Datasource` will rarely be the most convenient way class to use. Users will usually do better with `PandasDatasource`, `SparkDFDatasource`, `SqlAlchemyDatasource`, `SimpleSqlAlchemyDatasource`, etc.

abegong replied:
Hm. After reviewing, `PandasDatasource` and `SparkDFDatasource` are v2, not v3. I suspect that it would be relatively simple to create comparable classes in V3---which would save the user a lot of boilerplate configuration---but I haven't looked closely.

alexsherstinsky replied:
@Rachel-Reverie @talagluck @abegong @donaldheppner -- I agree with Abe in terms of what is v2 and what is v3.  In addition, I would like to reiterate that we need to allocate some development cycles to certify the use of `InferredAssetSqlDataConnector` and `ConfiguredAssetSqlDataConnector` (and make any updates to accomplish this certification) and the `Datasource` using these direct SQL data connectors.

As part of this effort, we will also document the recommended usage for `SimpleSqlalchemyDatasource`.  Today, all documented cases of the direct SQL data connectors are not certified, not used in CLI, while only `SimpleSqlalchemyDatasource` is officially certified (but appears in documentation a lot more than it should).

We also need to improve the way `Datasource` is created for and displayed in Jupyter notebooks generated by `CLI datasource` and `CLI suite` commands -- today, the nuance between `BatchRequest` and `RuntimeBatchRequest` is not handled at all (in addition to the `Datasource` not having the optimal data connector for the given CLI operation.

I was pretty sure that I created JIRA items for these issues before, but now, unfortunately, I am unable to find them.  Before re-entering them, perhaps someone with stronger Atlassian privileges could search for them.

Happy to discuss.  Thank you.

talagluck commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+#### Configuring your Datasource's Data Connectors
+
+Great Expectations provides three types of `DataConnector` classes, which are useful in various situations.  Which Data Connector you will want to use will depend on the format of your source data systems.
+- An InferredAssetDataConnector infers the `data_asset_name` by using a regex that takes advantage of patterns that exist in the filename or folder structure.  If your source data system is designed so that it can easily be parsed by regex, this will allow new data to be included by the Datasource automatically.
```
I see that this is the current definition in the How to Choose Which Data Connector to Use. That said, while this definition definitely works for data on a file system, you may want to confirm that it is accurate for databases as well.

Rachel-Reverie replied:
I've updated this section to include the `InferredAssetSqlDataConnector` for SQL based source data systems, and to specify that the `InferredAssetDataConnector` is used for filesystem based source data systems.

talagluck commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+#### Configuring your Datasource's Execution Engine
+
+After your Datasource's configuration has a `name` and `class_name` defined, you will need to define the `execution_engine`.  In your configuration the value of your `execution_engine` will at the very least contain the `class_name` of your Execution Engine, and may also include a `connection_string` if your source data system requires one.
```
```suggestion
After your Datasource's configuration has a `name` and `class_name` defined, you will need to define a single `execution_engine`.  In your configuration the value of your `execution_engine` will at the very least contain the `class_name` of your Execution Engine, and may also include a `connection_string` if your source data system requires one.
```

talagluck commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+In the `data_connectors` dictionary you may define multiple Data Connectors, including different types of Data Connectors, so long as they all have unique values in the place of the `<name_of_your_data_connector>` key.  We provide detailed guidance to help you decide on which Data Connectors to use in our guide: [How to choose which DataConnector to use](/docs/guides/connecting_to_your_data/how_t ...
+
+The `<additional_keys_based_on_source_data_system>` will be things like `base_directory` and `default_regex` for filesystems, or  `batch_identifiers` for SQL based data systems. For specifics on the additional keys that you can use in your Data Connectors' configurations, please see the corresponding guide for connecting to a specific source data system (since the keys you will need to define wil ...
```
I think `batch_identifiers` are a thing for filesystems, too. This is worth double checking.

abegong replied:
Yes, they should be.

Rachel-Reverie replied:
I've updated the documentation to indicate that.

talagluck commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+### 4. Save the Datasource configuration to your Data Context.
+
+What is the point of configuring a Datasource if you can't easily use it in the future?  It this point you will want to save your Datasource configuration to your Data Context.  This can be done easily by using the `add_datasource()` function, which is conveniently accessible from your Data Context.  
```
```suggestion
What is the point of configuring a Datasource if you can't easily use it in the future?  At this point you will want to save your Datasource configuration to your Data Context.  This can be done easily by using the `add_datasource()` function, which is conveniently accessible from your Data Context.  
```

talagluck reviewed (commented):
Looks great overall! Just a few minor corrections, and a few points that you may want to confirm.

abegong commented on docs/guides/connecting_to_your_data/overview.md:
````diff
@@ -0,0 +1,205 @@
+great_expectations datasource new
+```
+
+From there, you will be able to follow along a YAML based workflow for configuring and saving your Datasource.  Whether you prefer to work with the Jupyter Notebook's boilerplate for creating a datasource, or would rather dive in from scratch with a Python script, however, most of the work will take place in the configuring of the Datasource in question.
````
I think we're underselling `test_yaml_config` here.

I'll add a section in the "Core Promises for the “Connect to data : Overview” google doc that we've been collaborating in, since there are different versions of saying what we might want to say here.

abegong replied:
Okay, I've added a whole section at the bottom of the doc.

Rachel-Reverie replied:
I've extended this section with your additions from the google doc.

abegong commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+<!-- The following subsections should be repeated as necessary.  They should give a high level map of the things that need to be done or optionally can be done in this process, preferably in the order that they should be addressed (assuming there is one). If the process crosses multiple steps of the Universal Map, use the <SetupHeader> <ConnectHeader> <CreateHeader> and <ValidateHeader> tags to i ...
+
+### 1. Boilerplate
```
Is this the header we want?

Rachel-Reverie replied:
I'm not really committed to it.  Previously it was "Create your datasource" -- except we aren't really doing that through the CLI.  What we're doing is creating the boilerplate that will follow along with the steps after this.  I am open to renaming this section.

abegong commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,205 @@
+
+The `name` and `class_name` top level keys will be the first you need to define.  The `name` key can be anything you want, but it is best to use a descriptive name as you will use this to reference your Datasource in the future. Unless you are extending Great Expectations and using a subclass of Datasource, you will almost never need to use a `class_name` other than `Datasource` for the top level ...
+
+#### Configuring your Datasource's Execution Engine
```
I think this is too much detail for this guide---the details will depend on which type of DataSource you're using.

abegong reviewed (changes requested):
Good progress! My big question is about recommended workflow: if we want users to start with the CLI and `test_yaml_config` (I think that we do) and can assume that they're following that flow, then several parts of this doc could be simplified quite a bit.

The discussion about test_yaml_config is directly related to this question.

Inspecting the DataConnectors section, I've also realized that there's some (simple, I think?) work that we could do to make these workflows much easier for users, by subclassing Datasource for specific use cases (e.g. PandasDatasource). I'd like to talk with Don about this---maybe we could run these as experiments to streamline onboarding and setup?

This isn't feedback for this PR---more a separate thread to follow up on.

The section at the end on creating Batches needs quite a bit more. Unfortunately, our existing docs are not in great shape here. Let's sync up on how to flesh this out.

Last: we should figure out the picture for this article.

Rachel-Reverie committed:
Update docs/guides/connecting_to_your_data/overview.md

Co-authored-by: talagluck <tal@superconductive.com>

Rachel-Reverie committed:
Update docs/guides/connecting_to_your_data/overview.md

Co-authored-by: talagluck <tal@superconductive.com>

Rachel-Reverie committed:
Extended section on `test_yaml_config()`

Rachel-Reverie committed:
Merge remote-tracking branch 'origin/DOC-52_connect_to_data_overview' into DOC-52_connect_to_data_overview

Rachel-Reverie committed:
Removed sections of the ToC that were moved into "1. Setup" and "2. Connecting to Data".

Rachel-Reverie committed:
Added `batch_identifiers` as a potential key to replace <additional_keys_based_on_source_data_systems> in the examples when the source data system is a filesystem.

Rachel-Reverie committed:
Updated `InferredAssetDataConnector` section to include that `InferredAssetSqlDataConnector` provides similar functionality for SQL based source data systems.

Rachel-Reverie committed:
Added index for Step 2: Connect to Data to sidebar.js; updated contents of index to correspond to folder structure in sidebar.js

Rachel-Reverie committed:
Updated relative links for index.md

abegong commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,216 @@
+
+If something about your configuration wasn't set up correctly, `test_yamal_config()` will raise an error.  Whenever possible, `test_yamal_config()` provides helpful warnings and error messages.  It can't solve every problem, but it can solve many.
+
+You can call `test_yamal_config()` from your Data Context, like so:
```
```suggestion
You can call `test_yaml_config()` from your Data Context, like so:
```

abegong commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,216 @@
+- gathering a list of available DataAssets (e.g. tables in SQL; files or folders in a filesystem)
+- verifying that it can successfully fetch at least one Batch from the source.
+
+If something about your configuration wasn't set up correctly, `test_yamal_config()` will raise an error.  Whenever possible, `test_yamal_config()` provides helpful warnings and error messages.  It can't solve every problem, but it can solve many.
```
```suggestion
If something about your configuration wasn't set up correctly, `test_yaml_config()` will raise an error.  Whenever possible, `test_yaml_config()` provides helpful warnings and error messages.  It can't solve every problem, but it can solve many.
```

abegong commented on docs/guides/connecting_to_your_data/overview.md:
```diff
@@ -0,0 +1,216 @@
+
+### 3. Test your configuration
+
+Because the configurations for Datasources can vary depending on the underlying data system they are connecting to, Great Expectations provides a convenience function that will help you determine if there are errors in your configuration.  This function is `test_yaml_config()`.  Using `test_yamal_config` in a Jupyter Notebook is our recommended method for testing Datasource configuration.  Of cou ...
```
```suggestion
Because the configurations for Datasources can vary depending on the underlying data system they are connecting to, Great Expectations provides a convenience function that will help you determine if there are errors in your configuration.  This function is `test_yaml_config()`.  Using `test_yaml_config` in a Jupyter Notebook is our recommended method for testing Datasource configuration.  Of course, you can always edit and test YAML configs manually, and instantiate Datasources through code.  When executed, `test_yaml_config()` will instantiate the component and run through a self check procedure to verify the component works as expected.
```

Rachel-Reverie committed:
Update docs/guides/connecting_to_your_data/overview.md

Co-authored-by: Abe Gong <abegong@users.noreply.github.com>

Rachel-Reverie committed:
Update docs/guides/connecting_to_your_data/overview.md

Co-authored-by: Abe Gong <abegong@users.noreply.github.com>

Rachel-Reverie committed:
- Changed heading of section 1. Boilerplate to 1. Prepare scaffolding
- Refactored name of overview file to match naming convention of other step's overview files.
- Updated internal links and references to the refactored overview file's name.

Rachel-Reverie committed:
- Corrected spelling of `test_yaml_config()`

Rachel-Reverie committed:
Merge branch 'universal-map-docs-update' into DOC-52_connect_to_data_overview
